### ISSUE TITLE
Allow for some backwards carriage movement without going up to the next row.

### ISSUE THREAD

**Issue by Adrienne200** [AllYarnsAreBeautiful/ayab-desktop#244]
Before starting to knit across a row, sometimes you move the carriage backwards a bit. Can be because the yarn got caught in the brushes and you need to back up to untangle it, or in DBJ because the wrong yarn is in the color changer and you notice just in time as the carriage is meeting the first needle.

In Brother's native mode, they solve this issue by not counting up until after the carriage has passed the center needle. Which you hardly notice most of the time, but has its downsides, such as when knitting a large sideways triangle where you start at one side and keep increasing then decreasing. The carriage must pass the center even when you're only knitting near one end.
Page 79 of the KH940 manual "Whilst knitting in pattern you must take the K-carriage past the centre of the machine (Green 1 needle) on every row."  Also on page 43 of the 910 manual.

I wonder if AYAB could be a bit smarter than this such as allowing the countup so long as the carriage has passed say L25/R25.

I have had this happen a couple of times in real knitting, though it is not that easy to reproduce when trying to make it happen.  Possibly slightly more common in DBJ?  
Set up to knit a narrow piece in the center, and before starting across the needles move the carriage back and forth a few inches on the side of the bed. Sooner or later it will beep and has clicked up a row that you didn't actually knit.

**Comment by chris007de** [comment]
Currently, requesting the next line should only happen when the carriage was moved completely to the other side of the active area.
We'll investigate this.

**Comment by chris007de** [comment]
As this is currently discussed on Facebook (https://www.facebook.com/groups/1384431355220966/permalink/2059471511050277/?comment_id=2060182977645797), I looked into the code once more and found why this might be misbehaving currently:

This if-Statement handles when to request a new line: https://github.com/AllYarnsAreBeautiful/ayab-firmware/blob/master/knitter.cpp#L230
It requests a new line when the carriage leaves the active knitting area, plus some buffer ("END_OF_LINE_OFFSET_L/R). When entering this area again, the variable "_workedOnLine" is set to true again, which enables the request of the next line when leaving the knitting area + buffer on the other side again. Without having entered the knitting area + buffer, this variable gets not set to true, and without it being true, no new line is requested.
So when people report that a new line is requested while twiddling around the carriage, it must have already entered the active knitting area + buffer again and accidentially leaving it again on the same side.

Practical example:
- END_OF_LINE_OFFSET_L is 32, END_OF_LINE_OFFSET_R is 12
- Let's assume that Start needle is 25, stop needle is 25
- Carriage moved from left to right and passed needle 25+12 = 37 on the right, which triggers the software to request a new line
- The software will not request a new line unless entering and leaving the area 57<->37 again, either on the left side (which is intended, as we are moving from right to left for the next run), but currently it's not excluded to leave on the right side again. So if somebody stops pretty exact on or shortly after 37, the next line is requested, then moving the carriage a bit to 36 and again to 37 or 38, another next line gets requested.

Proposed fix is to set "_workedOnLine" to true only when the carriage is inside the active knitting area (without buffer), this yields a buffer of 32 and 12 needles for moving the carriage while fixing stuff.

**Comment by hamsterwoede** [comment]
I'm not that familiar with github, but is there a way to split this issue into two issues? Slight movement of the carriage is a completely different problem compared to a big movement (might it be by accident, or on purpose). A big movement/error could probably be fixed by putting a 'go back a line/panic'-button in the ayab desktop that you could push while knitting.

**Comment by chris007de** [comment]
@hamsterwoede Sure, just open a new issue and refer to the original one.

> A big movement/error could probably be fixed by putting a 'go back a line/panic'-button in the ayab desktop that you could push while knitting.

Afair, #227 already requests controls to choose the next line to be knit. I just entered a comment there with more background information why this may not be the solution for this (#244) issue (at least not in the current state of the protocol).

**Comment by hamsterwoede** [comment]
That makes sense. I don't think it would add any clarity to open a new issue. I agree that at the moment that it's probably not the right solution with the way the protocol is set up now.

**Comment by chris007de** [comment]
> Proposed fix is to set "_workedOnLine" to true only when the carriage is inside the active knitting area (without buffer), this yields a buffer of 32 and 12 needles for moving the carriage while fixing stuff.

Implemented in 4abd6cf4bf02baaa9627a8b69edcae6914d3b4c0, seems the behavior is much better now, no more additional new line requests.

**Comment by derchrisuk** [comment]
http://ayab-debug.s3-website.eu-central-1.amazonaws.com/

**Comment by Adrienne200** [comment]
Much better now, thank you. (Tested with post61 of 0.95) 
There is an extreme case where the issue can still happen. Say you are knitting a sleeve that starts wide at the armhole and decreases down to quite a narrow wrist. In AYAB you have to set up a rectangle that is the maximum width, and even down near the wrist you must take the carriage past the start/stop needles every row. While you are knitting the full width the carriage mechanics does not let you turn around inside the start/stop range which is why this fix works, but down near the wrist, once you are at least 8 stitches narrower on one side, it is possible to move the carriage backwards and have it click up when you didn't want it to. 
But I'm OK with it as it is now so I am closing the issue; there are other ideas being floated to allow for changing the width during knitting which would solve the sleeve-shaping case, ideas similar to #227 about changing the row number.
